// @rosen-clients/axios/dist/index.js
import { create as createAxiosInstance, } from '@rosen-clients/rate-limited-axios';
import JsonBigintFactory from 'json-bigint';
const JsonBigInt = JsonBigintFactory({
    alwaysParseAsBig: false,
    useNativeBigInt: true,
    constructorAction: 'ignore',
    protoAction: 'ignore',
});
const transformBigInt = (obj, bigIntObj, field) => {
    if (Array.isArray(obj)) {
        return obj.map((row, index) => {
            return transformBigInt(row, bigIntObj[index], field);
        });
    }
    if (!Object.hasOwn(obj, field.split('.')[0])) {
        return {
            ...obj,
        };
    }
    if (field.indexOf('.') !== -1) {
        const subKeys = field.split('.');
        return {
            ...obj,
            [subKeys[0]]: transformBigInt(obj[subKeys[0]], bigIntObj[subKeys[0]], subKeys.slice(1).join('.')),
        };
    }
    return { ...obj, [field]: BigInt(bigIntObj[field]) };
};
export const JsonFieldBigintFactory = (fields) => {
    return (data) => {
        try {
            let dataJson = JSON.parse(data);
            const dataBigInt = JsonBigInt.parse(data);
            fields.forEach((field) => {
                dataJson = transformBigInt(dataJson, dataBigInt, field);
            });
            return dataJson;
        }
        catch (e) {
            return data;
        }
    };
};
export const createAxiosInstanceWithHeaders = (url, headers) => {
    const instance = createAxiosInstance();
    instance.defaults.baseURL = url;
    Object.entries(headers).forEach((header) => {
        instance.defaults.headers.common[header[0]] = header[1];
    });
    return (config) => {
        return instance.request(config).then((response) => response.data);
    };
};
