// @rosen-clients/rate-limited-axios/dist/config.js
import { DummyLogger } from '@rosen-bridge/abstract-logger';
import { Semaphore } from 'await-semaphore';
export class RateLimitedAxiosConfig {
    static limitRules = [];
    static logger = new DummyLogger();
    /**
     * Add new rule to apply rate limit for it
     * @param pattern
     * @param rateLimit
     * @param throttleWindow
     */
    static addRule(pattern, rateLimit, throttleWindow, timeout) {
        RateLimitedAxiosConfig.removeRule(pattern);
        RateLimitedAxiosConfig.limitRules.push({
            pattern: new RegExp(pattern),
            semaphore: new Semaphore(rateLimit),
            throttleWindow: throttleWindow,
            timeout: timeout,
        });
    }
    /**
     * Remove rule from rate limited pattern list
     * @param pattern
     */
    static removeRule(pattern) {
        RateLimitedAxiosConfig.limitRules =
            RateLimitedAxiosConfig.limitRules.filter((r) => r.pattern.source !== pattern);
    }
    /**
     * Get current rate limit rules
     * @returns
     */
    static getRules() {
        return RateLimitedAxiosConfig.limitRules;
    }
    /**
     * Set logger for RateLimitedAxios
     * @param logger
     */
    static setLogger(logger) {
        RateLimitedAxiosConfig.logger = logger;
    }
    /**
     * get logger of RateLimitedAxios
     * @returns
     */
    static getLogger() {
        return RateLimitedAxiosConfig.logger;
    }
}
